tony-core/src/main/java/com/linkedin/tony/TaskMonitor.java

```diff
@@ -51,15 +93,77 @@ public void run() {
   }
 
   private void refreshMetrics() {
-    resourceCalculator.updateProcessTree();
-    refreshMaxMemoryBytes();
+    refreshMemoryBytesMetrics();
+    if (isGpuMachine) {
+      refreshGPUMetrics();
+    }
+    numRefreshes++;
   }
 
-  private void refreshMaxMemoryBytes() {
+  private void refreshMemoryBytesMetrics() {
+    resourceCalculator.updateProcessTree();
     double memoryBytes = resourceCalculator.getRssMemorySize();
-    if (memoryBytes > maxMemoryBytes.getValue()) {
-      maxMemoryBytes.setValue(memoryBytes);
-      metrics.setMetric(MAX_MEMORY_BYTES_INDEX, maxMemoryBytes);
+    setMaxMetrics(MAX_MEMORY_BYTES_INDEX, memoryBytes);
+    setAvgMetrics(AVG_MEMORY_BYTES_INDEX, memoryBytes);
+  }
+
+  private void refreshGPUMetrics() {
+    try {
+      GpuDeviceInformation gpuInfo = gpuDiscoverer.getGpuDeviceInformation();
+
+      double maxGpuUtilization = gpuInfo.getGpus().stream()
+          .mapToDouble(x -> x.getGpuUtilizations().getOverallGpuUtilization())
+          .max()
+          .getAsDouble();
+      double avgGpuUtilization = gpuInfo.getGpus().stream()
+          .mapToDouble(x -> x.getGpuUtilizations().getOverallGpuUtilization())
+          .average()
+          .getAsDouble();
+      double maxGpuMemoryUsage = gpuInfo.getGpus().stream()
+          .mapToDouble((x ->
+              ((double) x.getGpuMemoryUsage().getUsedMemoryMiB() / x.getGpuMemoryUsage().getTotalMemoryMiB() * 100)))
+          .max()
+          .getAsDouble();
+      double avgGpuMemoryUsage = gpuInfo.getGpus().stream()
+          .mapToDouble((x ->
+              ((double) x.getGpuMemoryUsage().getUsedMemoryMiB() / x.getGpuMemoryUsage().getTotalMemoryMiB() * 100)))
+          .average()
+          .getAsDouble();
+
+      setMaxMetrics(MAX_GPU_UTILIZATION_INDEX, maxGpuUtilization);
+      setAvgMetrics(AVG_GPU_UTILIZATION_INDEX, avgGpuUtilization);
+      setMaxMetrics(MAX_GPU_MEMORY_USAGE_INDEX, maxGpuMemoryUsage);
+      setAvgMetrics(AVG_GPU_MEMORY_USAGE_INDEX, avgGpuMemoryUsage);
+    } catch (GpuInfoException e) {
+      // Follow YARN's GPUDiscoverer mechanism of capping number of gpu metrics query
+      if (gpuDiscoverer.getNumOfErrorExecutionSinceLastSucceed() >= Constants.MAX_REPEATED_GPU_ERROR_ALLOWED) {
+        LOG.warn("Failed to collect GPU metrics", e);
+        isGpuMachine = false;
+        return;
+      }
+
+      LOG.warn("Failed to collect GPU metrics at " + (numRefreshes + 1) + "th run, retry...", e);
     }
   }
+
+  @VisibleForTesting
+  protected void setAvgMetrics(int metricIndex, double newMetricValue) {
```

Comment:
I think this and other `@VisibleForTesting` methods can be made package-private (no modifier) rather than `protected`.